File: Programming_Reference_problems/Leetcode/0100-0199/0109.Convert Sorted List to Binary Search Tree/Solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
struct ListNode* find(struct ListNode* start, struct ListNode* end) {
    struct ListNode* fast = start;
    struct ListNode* slow = start;
    while (fast != end && fast->next != end) {
        fast = fast->next->next;
        slow = slow->next;
    }
    return slow;
}

struct TreeNode* bulid(struct ListNode* start, struct ListNode* end) {
    if (start == end) {
        return NULL;
    }
    struct ListNode* node = find(start, end);
    struct TreeNode* ans = malloc(sizeof(struct TreeNode));
    ans->val = node->val;
    ans->left = bulid(start, node);
    ans->right = bulid(node->next, end);
    return ans;
}

struct TreeNode* sortedListToBST(struct ListNode* head) {
    return bulid(head, NULL);
}



/*
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 /*
struct TreeNode* helper(int arr[],int left, int right){
    if(left > right) return NULL;
    int mid = (left+right)/2;
    struct TreeNode* root = (struct TreeNode*)malloc(sizeof(struct TreeNode));
    root->val = arr[mid];
    root->left = helper(arr,left,mid-1);
    root->right = helper(arr,mid+1,right);
    return root;
}

struct TreeNode* sortedListToBST(struct ListNode* head) {
    if(!head) return NULL;
    if(!head->next){
        struct TreeNode* temp = (struct TreeNode*)malloc(sizeof(struct TreeNode));
        temp->val = head->val;
        temp->left = NULL; // C 에선 이런 초기화 필요
        temp->right = NULL;
        return temp;
    }
    int arr[100000];
    int idx = 0;
    while(head){
        arr[idx++] = head->val;
        head = head->next;
    }
    return helper(arr,0,idx-1);
}

*/

File: Programming_Reference_problems/Leetcode/0100-0199/0109.Convert Sorted List to Binary Search Tree/Solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        def buildBST(nums, start, end):
            if start > end:
                return None
            mid = (start + end) >> 1
            return TreeNode(
                nums[mid], buildBST(nums, start, mid - 1), buildBST(nums, mid + 1, end)
            )

        nums = []
        while head:
            nums.append(head.val)
            head = head.next
        return buildBST(nums, 0, len(nums) - 1)


# class Solution:
#     def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
#         if not head:
#             return None
#         if not head.next:
#             return TreeNode(head.val)

#         slow = head
#         fast = head
#         prev = None

#         # Finding the middle element
#         while fast and fast.next:
#             prev = slow
#             slow = slow.next
#             fast = fast.next.next

#         # Create the root node with the middle element
#         root = TreeNode(slow.val)

#         # Disconnect the left part of the list
#         if prev:
#             prev.next = None

#         # Recursively build left and right subtrees
#         root.left = self.sortedListToBST(head)    # Left subtree
#         root.right = self.sortedListToBST(slow.next)  # Right subtree

#         return root